CancellationToken support (#5203)
Particular/docs.particular.net: nservicebus/upgrades/7to8/index.md

* [Log4net](/nservicebus/logging/log4net.md)
* [NLog](/nservicebus/logging/nlog.md)

## CancellationToken support

NServiceBus version 8 supports cooperative cancellation using `CancellationToken`.

### Non-message-handling contexts

Methods on extension points are updated to include a mandatory `CancellationToken` parameter. This includes abstract classes and interfaces needed to implement a message transport or persistence libary, as well as other extension points like `IDataBus`, `FeatureStartupTask`, `INeedToInstallSomething`. Implementors can be updated by adding the `CancellationToken` parameter to the end of the method signature.

> Methods used outside the message processing pipeline now include an optional `CancellationToken` parameter, including methods for starting and stopping endpoints, and methods to Send/Publish messages from outside a message processing pipeline, such as from within a web application.

Reviewer: ```suggestion
Methods used outside the message processing pipeline now include an optional `CancellationToken` parameter, including methods for starting and stopping endpoints, and methods to send or publish messages from outside a message processing pipeline, such as from within a web application.
```

> It is recommended to forward a `CancellationToken` within a method to any method that accepts one. For example, within a web application controller:

Reviewer: ```suggestion
It is recommended to forward a `CancellationToken` to any method that accepts one. For example, within a web application controller:
```

> ```csharp

Reviewer 2: Should this be a snippet?

Author: Yeah all these snippets can be snippets, but I figured it was easier to just write it out while WIP and snippetize later.

public class TestController
{
    private IMessageSession session;

    public TestController(IMessageSession session)
    {
        this.session = session;
    }

    [HttpGet("/test")]
    public async Task<string> Get(CancellationToken cancellationToken)
    {
        // Forward the cancellation token to methods that accept one
        await Task.Delay(10_000, cancellationToken);

        return "Finished 10s delay";
    }
}
```

> [Enabling .NET source code analysis](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview) (enabled by default in projects targeting .NET 5 or above) is also recommended so that [CA2016: Forward the CancellationToken parameter to methods that take one](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2016) can identify locations where the cancellation token should be forwarded. This rule is presesented as an informational message only, but the [analyzer severity](https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers#configure-severity-levels) can be upgraded to a warning using an [.editorconfig file](https://editorconfig.org/):

Reviewer 2: ```suggestion
[Enabling .NET source code analysis](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview) (enabled by default in projects targeting .NET 5 or above) is also recommended so that [CA2016: Forward the CancellationToken parameter to methods that take one](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2016) can identify locations where the `CancellationToken` should be forwarded. This rule is presented as an informational message only, but the [analyzer severity](https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers#configure-severity-levels) can be upgraded to a warning using an [.editorconfig file](https://editorconfig.org/):
```

```ini
[*.cs]
dotnet_diagnostic.CA2016.severity = warning
```

### Message handlers and pipeline

Inside a message handler, a cancellation token from the incoming message processing pipeline is available on the `IMessageHandlerContext` as the property `context.CancellationToken`. The cancellation token was added to the context parameter to avoid making a breaking change to `IHandleMessages` affecting all message handlers and sagas.

Similarly, [pipeline behaviors](/nservicebus/pipeline/manipulate-with-behaviors.md) also contain a `CancellationToken` property on their respective `context` parameters.

Methods on `IMessageHandlerContext` such as `Send()` and `Publish()` do not accept a cancellation token, as the token from the incoming message pipeline will be routed to the outgoing operations transparently.

> It is recommended to forward the `context.CancellationToken` to any other method that accepts one. A new analyzer identifies locations where the token with a build warning, for example:

Reviewer: ```suggestion
It is recommended to forward the `context.CancellationToken` to any other method that accepts one. A new analyzer identifies locations where the token should be forwarded with a build warning, for example:
```

```csharp

> public class SampleHandler : IHandleMessages<TestMessage>

Reviewer 2: Snippet?

* [Log4net](/nservicebus/logging/log4net.md)
* [NLog](/nservicebus/logging/nlog.md)

## CancellationToken support

> NServiceBus version 8 supports [cooperative cancellation](/nservicebus/hosting/cooperative-cancellation.md) using `CancellationToken`. While the token is optional in message handlers and on `IMessageSession` methods on extension points are updated to include a mandatory `CancellationToken` parameter. This includes abstract classes and interfaces needed to implement a message transport or persistence libary, as well as other extension points like `IDataBus`, `FeatureStartupTask`, `INeedToInstallSomething`. Implementors can be updated by adding the `CancellationToken` parameter to the end of the method signature.

Reviewer: ```suggestion
NServiceBus version 8 supports [cooperative cancellation](/nservicebus/hosting/cooperative-cancellation.md) using `CancellationToken` parameters. Where appropriate, optional `CancellationToken` parameters have been added to public methods. This includes the abstract classes and interfaces required to implement a message transport or persistence library, and other extension points like `IDataBus`, `FeatureStartupTask`, and `INeedToInstallSomething`. Implementers can be updated by adding an optional `CancellationToken` parameter to the end of method signatures.
```

## Shutdown behavior

In all versions of NServiceBus, `endpoint.Stop()` immediately stops receiving new messages but waits for a period of time for currently running message handlers to complete.

> In NServiceBus version 7 and below, the MSMQ, RabbitMQ, and SqlServer transports allow handlers up to 30 seconds to complete before forcing an endpoint shutdown. In NServiceBus version 8, all transports block shutdown until all handlers complete.

Reviewer 2: I haven't tested but assume that they hang forever?

> The cancellation token available on the `IMessageHandlerContext` in NServiceBus version 8 is triggered when the host forces shutdown. For example, the .NET Generic Host signals the cancellation token after 5 seconds.

Reviewer: ```suggestion
The cancellation token available on the `IMessageHandlerContext` in NServiceBus version 8 is triggered when the host forces shutdown. For example, by default, the .NET Generic Host signals the cancellation token after 5 seconds.
```

> However, if handlers do not observe the cancellation token, they will complete before the endpoints shuts down. This means that if, for example, a handler calls `Task.Delay(TimeSpan.Infinite)`, the endpoint will never shut down.

Reviewer: ```suggestion
However, if handlers do not observe the cancellation token, they will be allowed to complete before the endpoints shuts down. This means that if, for example, a handler calls `Task.Delay(TimeSpan.Infinite)`, the endpoint will never shut down.
```

* [Log4net](/nservicebus/logging/log4net.md)
* [NLog](/nservicebus/logging/nlog.md)

> ## CancellationToken support

Reviewer 2: @DavidBoike @adamralph I wonder if we should move most of the content in this section to the new page I created and just make this a short paragraph linking to it?

In all versions of NServiceBus, `endpoint.Stop()` immediately stops receiving new messages but waits for a period of time for currently running message handlers to complete.

> In NServiceBus version 7 and below, the MSMQ, ASQ, RabbitMQ, and SqlServer transports allow handlers up to 30 seconds to complete before forcing an endpoint shutdown. In NServiceBus version 8, all transports block shutdown until all handlers complete.

Author: ```suggestion
In NServiceBus version 7 and below, the MSMQ, RabbitMQ, Azure Storage Queues, and SQL transports allow handlers up to 30 seconds to complete before forcing an endpoint shutdown. In NServiceBus version 8, all transports block shutdown until all handlers complete.
```
